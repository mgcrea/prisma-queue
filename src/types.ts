import type { Prisma, PrismaClient, QueueJob as PrismaQueueJob } from "@prisma/client";
import type { PrismaJob } from "./PrismaJob";

export type JobPayload = Prisma.InputJsonValue;
export type JobResult = Prisma.InputJsonValue;
export type DatabaseJob<T, U> = Omit<PrismaQueueJob, "payload" | "result"> & { payload: T; result: U };

export type JobCreator<T extends JobPayload> = (client: PrismaLightClient) => Promise<T>;
export type JobWorker<T extends JobPayload = JobPayload, U extends JobResult = JobResult> = (
  job: PrismaJob<T, U>,
  client: PrismaClient
) => Promise<U>;

export type PrismaLightClient = Omit<
  PrismaClient,
  "$connect" | "$disconnect" | "$on" | "$transaction" | "$use"
>;
