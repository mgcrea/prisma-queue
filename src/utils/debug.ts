import createDebug from 'debug';

export const debug = createDebug('prisma-queue');
