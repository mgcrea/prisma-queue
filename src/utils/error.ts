export const serializeError = (err: unknown) => {
  if (err instanceof Error) {
    return {
      name: err.name,
      message: err.message,
      stack: err.stack,
    };
  }
  return {
    name: "UnknownError",
    message: `${err}`,
    stack: undefined,
  };
};
