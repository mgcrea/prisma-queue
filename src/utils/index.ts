export * from './debug';
export * from './error';
export * from './prisma';
export * from './string';
export * from './time';
