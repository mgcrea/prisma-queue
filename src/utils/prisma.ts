import {Prisma} from '@prisma/client';
import assert from 'assert';

export const getTableName = (modelName: string): string => {
  const model = Prisma.dmmf.datamodel.models.find((model) => model.name === modelName);
  assert(model && model.dbName, `Did not foudn model=${modelName} in Prisma.dmmf!`);
  return model.dbName;
};
