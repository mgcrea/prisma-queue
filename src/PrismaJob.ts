import type { Prisma } from "@prisma/client";
import type { DatabaseJob, PrismaLightClient } from "./types";
import { isPrismaError } from "./utils";
// import { debug } from "./utils";

export type PrismaJobOptions = {
  model: Prisma.QueueJobDelegate;
  client: PrismaLightClient;
};

/**
 * Represents a job within a Prisma-managed queue.
 */
export class PrismaJob<Payload, Result> {
  #model: Prisma.QueueJobDelegate;
  #client: PrismaLightClient;
  #record: DatabaseJob<Payload, Result>;

  public readonly id;
  public readonly createdAt: Date = new Date();

  /**
   * Constructs a new PrismaJob instance with the provided job record and database access objects.
   * @param record - The initial database job record.
   * @param model - The Prisma delegate used for database operations related to the job.
   * @param client - The Prisma client for executing arbitrary queries.
   */
  constructor(record: DatabaseJob<Payload, Result>, { model, client }: PrismaJobOptions) {
    this.#model = model;
    this.#client = client;
    this.#record = record;
    this.id = record.id;
  }

  /**
   * Internal method to assign a new record to the job.
   * @param record - Optional new record to assign.
   */
  #assign(record?: DatabaseJob<Payload, Result>) {
    if (record) {
      this.#record = record;
    }
  }

  /**
   * Gets the current job record.
   */
  public get record(): DatabaseJob<Payload, Result> {
    return this.#record;
  }

  /**
   * Gets the job's unique key if any.
   */
  public get key() {
    return this.#record.key;
  }

  /**
   * Gets the job's queue name.
   */
  public get queue() {
    return this.#record.queue;
  }

  /**
   * Gets the CRON expression associated with the job for recurring scheduling.
   */
  public get cron() {
    return this.#record.cron;
  }

  /**
   * Gets the job's priority level.
   */
  public get priority() {
    return this.#record.priority;
  }

  /**
   * Gets the payload associated with the job.
   */
  public get payload() {
    return this.#record.payload;
  }

  /**
   * Gets the timestamp when the job was finished.
   */
  public get finishedAt() {
    return this.#record.finishedAt;
  }

  /**
   * Gets the error record if the job failed.
   */
  public get error() {
    return this.#record.error;
  }

  /**
   * Updates the job's progress percentage.
   * @param progress - The new progress percentage.
   */
  public async progress(progress: number): Promise<DatabaseJob<Payload, Result>> {
    return await this.update({ progress: Math.max(0, Math.min(100, progress)) });
  }

  /**
   * Fetches the latest job record from the database and updates the internal state.
   */
  public async fetch(): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.findUnique({
      where: { id: this.id },
    })) as DatabaseJob<Payload, Result>;
    this.#assign(record);
    return record;
  }

  /**
   * Updates the job record in the database with new data.
   * @param data - The new data to be merged with the existing job record.
   */
  public async update(data: Prisma.QueueJobUpdateInput): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.update({
      where: { id: this.id },
      data,
    })) as DatabaseJob<Payload, Result>;
    this.#assign(record);
    return record;
  }

  /**
   * Deletes the job from the database.
   */
  public async delete(): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.delete({
      where: { id: this.id },
    })) as DatabaseJob<Payload, Result>;
    return record;
  }

  /**
   * Checks if the job is currently locked by another transaction.
   * @returns {Promise<boolean>} True if the job is locked, false otherwise.
   */
  public async isLocked(): Promise<boolean> {
    try {
      // Attempt to select and lock the row with a timeout
      await this.#client.$executeRawUnsafe(
        `SELECT "id" FROM "public"."queue_jobs" WHERE "id" = $1 FOR UPDATE NOWAIT`,
        this.id,
      );

      // If we reach here, the row is not locked
      return false;
    } catch (error) {
      // Handle specific error types that indicate lock contention
      if (isPrismaError(error) && error.meta?.["code"] === "55P03") {
        // PostgreSQL's lock_not_available
        return true;
      }
      // Re-throw other unexpected errors
      throw error;
    }
  }
}
