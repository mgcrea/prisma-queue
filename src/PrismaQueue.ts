import {Prisma, PrismaClient} from '@prisma/client';
import assert from 'assert';
import Cron from 'croner';
import {EventEmitter} from 'events';
import {PrismaJob} from './PrismaJob';
import {DatabaseJob, JobCreator, JobPayload, JobResult, JobWorker} from './types';
import {calculateDelay, debug, escape, getTableName, serializeError, waitFor} from './utils';

export type PrismaQueueOptions<T, U> = {
  prisma?: PrismaClient;
  name?: string;
  maxConcurrency?: number;
  pollInterval?: number;
  // jobInterval?: number;
  tableName?: string;
};

export type EnqueueOptions = {
  runAt?: Date;
  key?: string;
  maxAttempts?: number;
  priority?: number;
};
export type ScheduleOptions = Omit<EnqueueOptions, 'key' | 'runAt'> & {
  key: string;
  cron: string;
};

const JOB_INTERVAL = 25;

export class PrismaQueue<T extends JobPayload = JobPayload, U extends JobResult = JobResult> extends EventEmitter {
  private name: string;
  private prisma: PrismaClient;
  private config: Required<Omit<PrismaQueueOptions<T, U>, 'name' | 'prisma'>>;

  private count = 0;
  private concurrency = 0;
  private stopped = true;

  public constructor(private options: PrismaQueueOptions<T, U> = {}, public worker?: JobWorker<T, U>) {
    super();

    const {
      prisma = new PrismaClient(),
      name = 'default',
      tableName = getTableName('QueueJob'),
      maxConcurrency = 1,
      pollInterval = 10000,
      // jobInterval = 100,
    } = options;

    assert(name.length <= 255, 'name must be less or equal to 255 chars');
    assert(pollInterval >= 100, 'pollInterval must be more than 100 ms');
    // assert(jobInterval >= 10, 'jobInterval must be more than 10 ms');

    this.name = name;
    this.prisma = prisma;
    this.config = {
      tableName,
      maxConcurrency,
      pollInterval,
      // jobInterval,
    };
  }

  private get model() {
    return this.prisma.queueJob;
  }

  public async start(): Promise<void> {
    this.stopped = false;
    return this.poll();
  }

  public async stop(): Promise<void> {
    this.stopped = true;
  }

  public async enqueue(payloadOrFunction: T | JobCreator<T>, options: EnqueueOptions = {}): Promise<PrismaJob<T, U>> {
    debug(`enqueue(${JSON.stringify(payloadOrFunction)})`);
    const {prisma, name: queueName} = this;
    const {key, maxAttempts, priority = 0, runAt} = options;
    const record = await prisma.$transaction(async (client) => {
      const payload = payloadOrFunction instanceof Function ? await payloadOrFunction(client) : payloadOrFunction;
      const data = {queue: queueName, payload, maxAttempts, priority, runAt};
      if (key && runAt) {
        const {count} = await this.model.deleteMany({
          where: {
            key,
            runAt: {
              gte: new Date(),
              not: runAt,
            },
          },
        });
        if (count > 0) {
          debug(`deleted ${count} conflicting upcoming queue jobs`);
        }
        return await this.model.upsert({
          where: {key_runAt: {key, runAt}},
          create: {key, ...data},
          update: data,
        });
      }
      return await this.model.create({data});
    });
    return new PrismaJob(record as DatabaseJob<T, U>, {prisma});
  }

  public async schedule(options: ScheduleOptions, payloadOrFunction: T | JobCreator<T>): Promise<PrismaJob<T, U>> {
    debug(`schedule(${JSON.stringify(options)})`);
    const {key, cron, ...otherOptions} = options;
    const runAt = Cron(cron).next();
    assert(runAt, `Failed to find a future occurence for given cron`);
    return this.enqueue(payloadOrFunction, {key, runAt, ...otherOptions});
  }

  private async poll(): Promise<void> {
    if (this.stopped) {
      return;
    }
    const {maxConcurrency, pollInterval} = this.config;
    let estimatedQueueSize = await this.size();
    while (estimatedQueueSize > 0) {
      while (estimatedQueueSize > 0 && this.concurrency < maxConcurrency) {
        estimatedQueueSize--;
        this.concurrency++;
        setImmediate(() =>
          this.dequeue()
            .then((count) => {
              if (count === 0) {
                estimatedQueueSize = 0;
              }
            })
            .catch((err) => this.emit('error', err))
            .finally(() => this.concurrency--)
        );
      }

      await waitFor(JOB_INTERVAL);
    }

    debug(`poll waiting for pollInterval=${pollInterval}`);
    await waitFor(pollInterval);

    await this.poll();
  }

  // https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/
  // @TODO https://docs.quirrel.dev/api/queue
  // https://github.com/quirrel-dev/quirrel/blob/main/package.json
  private async dequeue(): Promise<number | undefined> {
    debug(`dequeue()`);
    if (this.stopped) {
      return;
    }
    const {prisma, name: queueName} = this;
    const tableName = escape(this.config.tableName);
    return await prisma.$transaction(
      async (client) => {
        const rows = await client.$queryRawUnsafe<DatabaseJob<T, U>[]>(
          `UPDATE ${tableName} SET "processedAt" = NOW(), "attempts" = "attempts" + 1
           WHERE id = (
             SELECT id
             FROM ${tableName}
             WHERE (${tableName}."queue" = $1)
               AND (${tableName}."runAt" < NOW())
               AND (${tableName}."finishedAt" IS NULL)
               AND (${tableName}."notBefore" IS NULL OR ${tableName}."notBefore" < NOW())
             ORDER BY ${tableName}."priority" ASC, ${tableName}."runAt" ASC
             FOR UPDATE SKIP LOCKED
             LIMIT 1
           )
           RETURNING *;`,
          queueName
        );
        // d({ rows });
        if (rows.length < 1) {
          return 0;
        }
        const {id, payload, attempts} = rows[0];
        const job = new PrismaJob(rows[0], {prisma: client});
        let result;
        try {
          assert(this.worker, 'Missing queue worker to process job');
          debug(`starting worker for job({id: ${id}, payload: ${JSON.stringify(payload)}})`);
          result = await this.worker(job, prisma);
          await client.queueJob.update({
            where: {id: id},
            data: {finishedAt: new Date(), progress: 100, result, error: Prisma.DbNull},
          });
        } catch (err) {
          debug(`failed finishing job({id: ${id}, payload: ${JSON.stringify(payload)}}) with error="${err}"`);
          const delay = calculateDelay(attempts);
          const nextRunAt = new Date(Date.now() + delay);
          debug(`will retry at runAt=${nextRunAt.toISOString()} (attempts=${attempts})`);
          await client.queueJob.update({
            where: {id: id},
            data: {
              failedAt: new Date(),
              error: serializeError(err),
              notBefore: new Date(Date.now() + calculateDelay(attempts)),
            },
          });
        }
        return rows.length;
      },
      // @NOTE https://github.com/prisma/prisma/issues/11565#issuecomment-1031380271
      {timeout: 864e5}
    );
  }

  public async size(): Promise<number> {
    const {prisma, name: queueName} = this;
    return await prisma.queueJob.count({where: {queue: queueName, finishedAt: null}});
  }
}
