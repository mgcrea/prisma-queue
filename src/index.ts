import { PrismaQueue, PrismaQueueOptions } from "./PrismaQueue";
import type { JobPayload, JobResult, JobWorker } from "./types";

export * from "./PrismaJob";
export * from "./PrismaQueue";

export const createQueue = <T extends JobPayload = JobPayload, U extends JobResult = JobResult>(
  options: PrismaQueueOptions,
  worker?: JobWorker<T, U>
) => {
  return new PrismaQueue<T, U>(options, worker);
};
