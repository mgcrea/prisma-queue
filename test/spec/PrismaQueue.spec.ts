import { PrismaClient } from "@prisma/client";
import type { PrismaQueue } from "src/index";
import { PrismaJob } from "src/PrismaJob";
import { debug, serializeError } from "src/utils";
import {
  createEmailQueue,
  prisma,
  waitForNextEvent,
  waitForNextJob,
  waitForNthJob,
  type JobPayload,
  type JobResult,
} from "test/utils";
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";

describe("PrismaQueue", () => {
  it("should properly create a queue", () => {
    const emailQueue = createEmailQueue();
    expect(emailQueue).toBeDefined();
    expect(Object.keys(emailQueue)).toMatchSnapshot();
  });
  describe("enqueue", () => {
    let queue: PrismaQueue<JobPayload, JobResult>;
    beforeAll(async () => {
      queue = createEmailQueue();
    });
    beforeEach(async () => {
      await prisma.queueJob.deleteMany();
      queue.start();
    });
    afterEach(async () => {
      queue.stop();
    });
    it("should properly enqueue a job", async () => {
      const job = await queue.enqueue({ email: "foo@bar.com" });
      expect(job).toBeInstanceOf(PrismaJob);
      expect(Object.keys(job)).toMatchSnapshot();
      const record = await job.fetch();
      expect(record?.payload).toEqual({ email: "foo@bar.com" });
      expect(record?.runAt).toBeInstanceOf(Date);
    });
  });

  describe("schedule", () => {
    let queue: PrismaQueue<JobPayload, JobResult>;
    beforeAll(async () => {
      queue = createEmailQueue();
    });
    beforeEach(async () => {
      await prisma.queueJob.deleteMany();
      queue.start();
    });
    afterEach(async () => {
      queue.stop();
    });
    it("should properly schedule a recurring job", async () => {
      const job = await queue.schedule(
        { key: "email-schedule", cron: "5 5 * * *" },
        { email: "foo@bar.com" }
      );
      expect(job).toBeInstanceOf(PrismaJob);
      const record = await job.fetch();
      expect(record).toBeDefined();
      expect(record?.runAt.getHours()).toBe(5);
      expect(record?.runAt.getMinutes()).toBe(5);
    });
    it("should properly re-enqueue a recurring job", async () => {
      await queue.schedule(
        { key: "email-schedule", cron: "5 5 * * *", runAt: new Date() },
        { email: "foo@bar.com" }
      );
      queue.start();
      await waitForNextEvent(queue, "enqueue");
      const jobs = await prisma.queueJob.findMany({ where: { key: "email-schedule" } });
      expect(jobs.length).toBe(2);
      const record = jobs[1];
      expect(record).toBeDefined();
      expect(record?.runAt.getHours()).toBe(5);
      expect(record?.runAt.getMinutes()).toBe(5);
    });
    it("should properly upsert a recurring job", async () => {
      await queue.schedule({ key: "email-schedule", cron: "5 5 * * *" }, { email: "foo@bar.com" });
      await queue.schedule({ key: "email-schedule", cron: "5 5 * * *" }, { email: "baz@bar.com" });
      const jobs = await prisma.queueJob.findMany({ where: { key: "email-schedule" } });
      expect(jobs.length).toBe(1);
      expect(jobs[0]?.payload).toEqual({ email: "baz@bar.com" });
    });
    it("should properly upsert a recurring job with another schedule", async () => {
      await queue.schedule({ key: "email-schedule", cron: "5 5 * * *" }, { email: "foo@bar.com" });
      await queue.schedule({ key: "email-schedule", cron: "0 5 * * *" }, { email: "foo@bar.com" });
      const jobs = await prisma.queueJob.findMany({ where: { key: "email-schedule" } });
      expect(jobs.length).toBe(1);
      expect(jobs[0]?.runAt.getMinutes()).toBe(0);
    });
  });

  describe("dequeue", () => {
    let queue: PrismaQueue<JobPayload, JobResult>;
    beforeAll(async () => {
      queue = createEmailQueue();
    });
    beforeEach(async () => {
      await prisma.queueJob.deleteMany();
      queue.start();
    });
    afterEach(async () => {
      queue.stop();
    });
    it("should properly dequeue a successful job", async () => {
      queue.worker = vi.fn(async (_job) => {
        return { code: "200" };
      });
      const job = await queue.enqueue({ email: "foo@bar.com" });
      await waitForNextJob(queue);
      expect(queue.worker).toHaveBeenCalledTimes(1);
      expect(queue.worker).toHaveBeenNthCalledWith(1, expect.any(PrismaJob), expect.any(PrismaClient));
      const record = await job.fetch();
      expect(record?.finishedAt).toBeInstanceOf(Date);
    });
    it("should properly dequeue a failed job", async () => {
      let error: Error | null = null;
      queue.worker = vi.fn(async (_job) => {
        error = new Error("failed");
        throw error;
      });
      const job = await queue.enqueue({ email: "foo@bar.com" });
      await waitForNextJob(queue);
      expect(queue.worker).toHaveBeenCalledTimes(1);
      expect(queue.worker).toHaveBeenNthCalledWith(1, expect.any(PrismaJob), expect.any(PrismaClient));
      const record = await job.fetch();
      expect(record?.finishedAt).toBeNull();
      expect(record?.error).toEqual(serializeError(error));
    });
    afterAll(() => {
      queue.stop();
    });
  });

  describe("priority", () => {
    let queue: PrismaQueue<JobPayload, JobResult>;
    beforeAll(async () => {
      queue = createEmailQueue();
    });
    beforeEach(async () => {
      await prisma.queueJob.deleteMany();
      // queue.start();
    });
    afterEach(async () => {
      queue.stop();
    });
    it("should properly prioritize a job with a lower priority", async () => {
      queue.worker = vi.fn(async (_job) => {
        return { code: "200" };
      });
      await queue.enqueue({ email: "foo@bar.com" });
      await queue.enqueue({ email: "baz@bar.com" }, { priority: -1 });
      queue.start();
      await waitForNthJob(queue, 2);
      expect(queue.worker).toHaveBeenCalledTimes(2);
      expect(queue.worker).toHaveBeenNthCalledWith(
        1,
        expect.objectContaining({
          payload: { email: "baz@bar.com" },
        }),
        expect.any(PrismaClient)
      );
      expect(queue.worker).toHaveBeenNthCalledWith(
        2,
        expect.objectContaining({
          payload: { email: "foo@bar.com" },
        }),
        expect.any(PrismaClient)
      );
    });
    afterAll(() => {
      queue.stop();
    });
  });

  describe("Job.progress()", () => {
    let queue: PrismaQueue<JobPayload, JobResult>;
    beforeAll(async () => {
      queue = createEmailQueue();
    });
    beforeEach(async () => {
      await prisma.queueJob.deleteMany();
      queue.start();
    });
    afterEach(async () => {
      queue.stop();
    });
    it("should properly update job progress", async () => {
      queue.worker = vi.fn(async (job) => {
        debug("working...", job.id, job.payload);
        await job.progress(50);
        throw new Error("failed");
      });
      const job = await queue.enqueue({ email: "foo@bar.com" });
      queue.start();
      await waitForNextJob(queue);
      const record = await job.fetch();
      expect(record?.progress).toBe(50);
    });
    afterAll(() => {
      queue.stop();
    });
  });
});
