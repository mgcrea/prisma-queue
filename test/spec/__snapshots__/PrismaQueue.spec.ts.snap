// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrismaQueue > enqueue > should properly enqueue a job 1`] = `
[
  "id",
]
`;

exports[`PrismaQueue > should properly create a queue 1`] = `
[
  "_events",
  "_eventsCount",
  "_maxListeners",
  "name",
  "config",
  "concurrency",
  "stopped",
  "options",
  "worker",
]
`;
